### ISSUE TITLE
[PATCH] debugger prints '1' when an reference or object is thrown

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16217]
Migrated from [rt.perl.org#132383](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=132383) (status was 'resolved')

Searchable as RT132383$

**Comment by p5pRT** [comment]
### From jkahrman@mathworks.com

This is a bug report for perl from jkahrman@&#8203;mathworks\.com\,
generated with the help of perlbug 1\.40 running under perl 5\.20\.2\.

If an exception object is thrown while stepping through the debugger\, the
debugger prints '1 at perl5db\.pl line 8616' instead of stringifing the object
and reporting the correct line number the exception was thrown \(See attachment
exception\_bug\)\.

The same behavior can be demonstrated more simply with unblessed references&#8203;:


---
% perl \-e 'die\(\{\}\);' || echo
HASH\(0xc93e78\)

% perl \-de 'die\(\{\}\);' || echo

Loading DB routines from perl5db\.pl version 1\.44
Editor support available\.

Enter h or 'h h' for help\, or 'man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:    die\(\{\}\);
&nbsp;  DB\<1> c
1 at /usr/share/perl/5\.20/perl5db\.pl line 8616\.
Debugged program terminated\.  Use q to quit or R to restart\,
use o inhibit\_exit to avoid stopping after program termination\,
h q\, h R or h o to get additional info\.
&nbsp;  DB\<1> q


---
This behavior is because perl5db\.pl is calling Carp&#8203;::longmess in scalar context
instead of list context\. When the first argument of Carp&#8203;::longmess \(or
shortmess\) is a reference\, it returns it's argument list \(@&#8203;\_\) unmodified\.

I've attached a patch \(debug\_exception\_1\.patch\) that calls Carp&#8203;::longmess in
list context to provide more useful messages\.


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low
&nbsp;    Type=Patch
&nbsp;    PatchStatus=HasPatch


---
Site configuration information for perl 5\.20\.2&#8203;:

Configured by Debian Project at Mon Jan 11 23&#8203;:24&#8203;:04 UTC 2016\.

Summary of my perl5 \(revision 5 version 20 subversion 2\) configuration&#8203;:
&nbsp;   
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.16\.0\-4\-amd64\, archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux themisto 3\.16\.0\-4\-amd64 \#1 smp debian 3\.16\.7\-ckt20\-1\+deb8u1 \(2015\-12\-14\) x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-D\_FORTIFY\_SOURCE=2 \-g \-O2 \-fstack\-protector\-strong \-Wformat \-Werror=format\-security \-Dldflags= \-Wl\,\-z\,relro \-Dlddlflags=\-shared \-Wl\,\-z\,relro \-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.20 \-Darchlib=/usr/lib/x86\_64\-linux\-gnu/perl/5\.20 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/x86\_64\-linux\-gnu/perl5/5\.20 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.20\.2 \-Dsitearch=/usr/local/lib/x86\_64\-linux\-gnu/perl/5\.20\.2 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Duse64bitint \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-Ui\_libutil \-Uversiononly \-DDEBUGGING=\-g \-Doptimize=\-O2 \-Duseshrplib \-Dlibperl=libperl\.so\.5\.20\.2 \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fwrapv \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fwrapv \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.9\.2'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/4\.9/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=libc\-2\.19\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.20
&nbsp;    gnulibc\_version='2\.19'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    DEBPKG&#8203;:debian/cpan\_definstalldirs \- Provide a sensible INSTALLDIRS default for modules installed from CPAN\.
&nbsp;    DEBPKG&#8203;:debian/db\_file\_ver \- http&#8203;://bugs\.debian\.org/340047 Remove overly restrictive DB\_File version check\.
&nbsp;    DEBPKG&#8203;:debian/doc\_info \- Replace generic man\(1\) instructions with Debian\-specific information\.
&nbsp;    DEBPKG&#8203;:debian/enc2xs\_inc \- http&#8203;://bugs\.debian\.org/290336 Tweak enc2xs to follow symlinks and ignore missing @&#8203;INC directories\.
&nbsp;    DEBPKG&#8203;:debian/errno\_ver \- http&#8203;://bugs\.debian\.org/343351 Remove Errno version check due to upgrade problems with long\-running processes\.
&nbsp;    DEBPKG&#8203;:debian/libperl\_embed\_doc \- http&#8203;://bugs\.debian\.org/186778 Note that libperl\-dev package is required for embedded linking
&nbsp;    DEBPKG&#8203;:fixes/respect\_umask \- Respect umask during installation
&nbsp;    DEBPKG&#8203;:debian/writable\_site\_dirs \- Set umask approproately for site install directories
&nbsp;    DEBPKG&#8203;:debian/extutils\_set\_libperl\_path \- EU&#8203;:MM&#8203;: set location of libperl\.a under /usr/lib
&nbsp;    DEBPKG&#8203;:debian/no\_packlist\_perllocal \- Don't install \.packlist or perllocal\.pod for perl or vendor
&nbsp;    DEBPKG&#8203;:debian/prefix\_changes \- Fiddle with \*PREFIX and variables written to the makefile
&nbsp;    DEBPKG&#8203;:debian/fakeroot \- Postpone LD\_LIBRARY\_PATH evaluation to the binary targets\.
&nbsp;    DEBPKG&#8203;:debian/instmodsh\_doc \- Debian policy doesn't install \.packlist files for core or vendor\.
&nbsp;    DEBPKG&#8203;:debian/ld\_run\_path \- Remove standard libs from LD\_RUN\_PATH as per Debian policy\.
&nbsp;    DEBPKG&#8203;:debian/libnet\_config\_path \- Set location of libnet\.cfg to /etc/perl/Net as /usr may not be writable\.
&nbsp;    DEBPKG&#8203;:debian/mod\_paths \- Tweak @&#8203;INC ordering for Debian
&nbsp;    DEBPKG&#8203;:debian/module\_build\_man\_extensions \- http&#8203;://bugs\.debian\.org/479460 Adjust Module&#8203;::Build manual page extensions for the Debian Perl policy
&nbsp;    DEBPKG&#8203;:debian/prune\_libs \- http&#8203;://bugs\.debian\.org/128355 Prune the list of libraries wanted to what we actually need\.
&nbsp;    DEBPKG&#8203;:fixes/net\_smtp\_docs \- \[rt\.cpan\.org \#36038\] http&#8203;://bugs\.debian\.org/100195 Document the Net&#8203;::SMTP 'Port' option
&nbsp;    DEBPKG&#8203;:debian/perlivp \- http&#8203;://bugs\.debian\.org/510895 Make perlivp skip include directories in /usr/local
&nbsp;    DEBPKG&#8203;:debian/deprecate\-with\-apt \- http&#8203;://bugs\.debian\.org/747628 Point users to Debian packages of deprecated core modules
&nbsp;    DEBPKG&#8203;:debian/squelch\-locale\-warnings \- http&#8203;://bugs\.debian\.org/508764 Squelch locale warnings in Debian package maintainer scripts
&nbsp;    DEBPKG&#8203;:debian/skip\-upstream\-git\-tests \- Skip tests specific to the upstream Git repository
&nbsp;    DEBPKG&#8203;:debian/patchlevel \- http&#8203;://bugs\.debian\.org/567489 List packaged patches for 5\.20\.2\-3\+deb8u3 in patchlevel\.h
&nbsp;    DEBPKG&#8203;:debian/skip\-kfreebsd\-crash \- http&#8203;://bugs\.debian\.org/628493 \[perl \#96272\] Skip a crashing test case in t/op/threads\.t on GNU/kFreeBSD
&nbsp;    DEBPKG&#8203;:fixes/document\_makemaker\_ccflags \- http&#8203;://bugs\.debian\.org/628522 \[rt\.cpan\.org \#68613\] Document that CCFLAGS should include $Config\{ccflags\}
&nbsp;    DEBPKG&#8203;:debian/find\_html2text \- http&#8203;://bugs\.debian\.org/640479 Configure CPAN&#8203;::Distribution with correct name of html2text
&nbsp;    DEBPKG&#8203;:debian/perl5db\-x\-terminal\-emulator\.patch \- http&#8203;://bugs\.debian\.org/668490 Invoke x\-terminal\-emulator rather than xterm in perl5db\.pl
&nbsp;    DEBPKG&#8203;:debian/cpan\-missing\-site\-dirs \- http&#8203;://bugs\.debian\.org/688842 Fix CPAN&#8203;::FirstTime defaults with nonexisting site dirs if a parent is writable
&nbsp;    DEBPKG&#8203;:fixes/memoize\_storable\_nstore \- \[rt\.cpan\.org \#77790\] http&#8203;://bugs\.debian\.org/587650 Memoize&#8203;::Storable&#8203;: respect 'nstore' option not respected
&nbsp;    DEBPKG&#8203;:debian/regen\-skip \- Skip a regeneration check in unrelated git repositories
&nbsp;    DEBPKG&#8203;:fixes/regcomp\-mips\-optim \- \[perl \#122817\] http&#8203;://bugs\.debian\.org/754054 Downgrade the optimization of regcomp\.c on mips and mipsel due to a gcc\-4\.9 bug
&nbsp;    DEBPKG&#8203;:debian/makemaker\-pasthru \- http&#8203;://bugs\.debian\.org/758471 Pass LD settings through to subdirectories
&nbsp;    DEBPKG&#8203;:fixes/perldoc\-less\-R \- \[rt\.cpan\.org \#98636\] http&#8203;://bugs\.debian\.org/758689 Tell the 'less' pager to allow terminal escape sequences
&nbsp;    DEBPKG&#8203;:fixes/pod\_man\_reproducible\_date \- http&#8203;://bugs\.debian\.org/759405 Support POD\_MAN\_DATE in Pod&#8203;::Man for the left\-hand footer
&nbsp;    DEBPKG&#8203;:fixes/io\_uncompress\_gunzip\_inmemory \- http&#8203;://bugs\.debian\.org/747363 \[rt\.cpan\.org \#95494\] Fix gunzip to in\-memory file handle
&nbsp;    DEBPKG&#8203;:fixes/socket\_test\_recv\_fix \- http&#8203;://bugs\.debian\.org/758718 \[perl \#122657\] Compare recv return value to peername in socket test
&nbsp;    DEBPKG&#8203;:fixes/hurd\_socket\_recv\_todo \- http&#8203;://bugs\.debian\.org/758718 \[perl \#122657\] TODO checking the result of recv\(\) on hurd
&nbsp;    DEBPKG&#8203;:fixes/regexp\-performance \- \[0fa70a0\] http&#8203;://bugs\.debian\.org/777556 \[perl \#123743\] simpify and speed up /\.\*\.\.\./ handling
&nbsp;    DEBPKG&#8203;:fixes/failed\_require\_diagnostics \- http&#8203;://bugs\.debian\.org/781120 \[perl \#123270\] Report inaccesible file on failed require
&nbsp;    DEBPKG&#8203;:fixes/array\-cloning \- http&#8203;://bugs\.debian\.org/779357 \[perl \#124127\] \[902d169\] fix cloning arrays with unused elements
&nbsp;    DEBPKG&#8203;:fixes/perldb\-threads \- http&#8203;://bugs\.debian\.org/779357 \[perl \#124127\] \[41ef2c6\] lib/perl5db\.pl&#8203;: Restore noop lock prototype
&nbsp;    DEBPKG&#8203;:fixes/CVE\-2015\-8607\_file\_spec\_taint\_fix \- ensure File&#8203;::Spec&#8203;::canonpath\(\) preserves taint
&nbsp;    DEBPKG&#8203;:fixes/encode\-unicode\-bom \- http&#8203;://bugs\.debian\.org/798727 \[rt\.cpan\.org \#107043\] Address https://rt.cpan.org/Public/Bug/Display.html?id=107043
&nbsp;    DEBPKG&#8203;:debian/encode\-unicode\-bom\-doc \- http&#8203;://bugs\.debian\.org/798727 Document Debian backport of Encode&#8203;::Unicode fix



---
@&#8203;INC for perl 5\.20\.2&#8203;:
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/x86\_64\-linux\-gnu/perl/5\.20\.2
&nbsp;    /usr/local/share/perl/5\.20\.2
&nbsp;    /usr/lib/x86\_64\-linux\-gnu/perl5/5\.20
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/x86\_64\-linux\-gnu/perl/5\.20
&nbsp;    /usr/share/perl/5\.20
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.



---
Environment for perl 5\.20\.2&#8203;:
&nbsp;    HOME=/home/jkahrman
&nbsp;    LANG=en\_US\.utf8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH=
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/mathworks/devel/sandbox/jkahrman/svn/1\.5\.1/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/sbin&#8203;:/usr/local/netbin&#8203;:/hub/bat/share/bin&#8203;:/hub/share/bin&#8203;:/hub/bat/share/p5
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/usr/local/bin/tcsh

**Comment by p5pRT** [comment]
### From jkahrman@mathworks.com

<details><summary>exception_bug</summary>

```diff
% perl -MError -e 'Error::Simple->new('Hello')->throw;' 
Hello at -e line 1.
% perl -MError -de 'Error::Simple->new('Hello')->throw;'

Loading DB routines from perl5db.pl version 1.44
Editor support available.

Enter h or 'h h' for help, or 'man perldebug' for more help.

main::(-e:1):	Error::Simple->new(Hello)->throw;
  DB<1> c
1 at /usr/share/perl/5.20/perl5db.pl line 8616.
Debugged program terminated.  Use q to quit or R to restart,
use o inhibit_exit to avoid stopping after program termination,
h q, h R or h o to get additional info.
  DB<1> q

```
</details>

**Comment by p5pRT** [comment]
### From jkahrman@mathworks.com

<details><summary>debug_exception_1.patch</summary>

```diff
--- perl5.20.2/lib/perl5db.pl	2017-02-06 20:00:13.000000000 -0500
+++ perl5.20.2_db_patch/lib/perl5db.pl	2017-02-06 20:00:13.000000000 -0500
@@ -8553,7 +8553,7 @@

     # We can call Carp::longmess without its being "debugged" (which we
     # don't want - we just want to use it!). Capture this for later.
-    my $mess = Carp::longmess(@_);
+    my ($mess) = Carp::longmess(@_);

     # Restore $single and $trace to their original values.
     ( $single, $trace ) = ( $mysingle, $mytrace );
@@ -8610,7 +8610,7 @@
     {

         package Carp;    # Do not include us in the list
-        eval { $mess = Carp::longmess(@_); };
+        eval { ($mess) = Carp::longmess(@_); };
     }
     ( $single, $trace ) = ( $mysingle, $mytrace );
     die $mess;
--- perl5.20.2/lib/perl5db.t	2017-02-06 20:00:13.000000000 -0500
+++ perl5.20.2_db_patch/lib/perl5db.t	2017-02-06 20:00:13.000000000 -0500
@@ -29,7 +29,7 @@
     $ENV{PERL_RL} = 'Perl'; # Suppress system Term::ReadLine::Gnu
 }

-plan(125);
+plan(126);

 my $rc_filename = '.perldb';

@@ -2357,6 +2357,24 @@
     );
 }

+# Test message from exception caught by debugger
+{
+    my $wrapper = DebugWrap->new(
+        {
+            cmds =>
+            [
+                'c',
+                'q',
+            ],
+            prog => '../lib/perl5db/t/ref_warn_die',
+        }
+    );
+
+    $wrapper->output_like(qr{^\QHASH(0x\E\w+\)},
+        'Test message from exception caught by debugger',
+    );
+}
+
 # Test the dieLevel option
 {
     my $wrapper = DebugWrap->new(
--- perl5.20.2/lib/perl5db/t/ref_warn_die	1969-12-31 19:00:00.000000000 -0500
+++ perl5.20.2_db_patch/lib/perl5db/t/ref_warn_die	2017-10-31 16:40:43.938018834 -0400
@@ -0,0 +1,5 @@
+#!*perl*
+
+warn([]);
+
+die({});

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed\, 01 Nov 2017 01&#8203;:03&#8203;:53 GMT\, jkahrman@&#8203;mathworks\.com wrote&#8203;:
> This is a bug report for perl from jkahrman@&#8203;mathworks\.com\,
> generated with the help of perlbug 1\.40 running under perl 5\.20\.2\.
> 
> If an exception object is thrown while stepping through the debugger\,
> the
> debugger prints '1 at perl5db\.pl line 8616' instead of stringifing the
> object
> and reporting the correct line number the exception was thrown \(See
> attachment
> exception\_bug\)\.
> 
> The same behavior can be demonstrated more simply with unblessed
> references&#8203;:
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> % perl \-e 'die\(\{\}\);' || echo
> HASH\(0xc93e78\)
> 
> % perl \-de 'die\(\{\}\);' || echo
> 
> Loading DB routines from perl5db\.pl version 1\.44
> Editor support available\.
> 
> Enter h or 'h h' for help\, or 'man perldebug' for more help\.
> 
> main&#8203;::\(\-e&#8203;:1\)&#8203;:   die\(\{\}\);
>   DB\<1> c
> 1 at /usr/share/perl/5\.20/perl5db\.pl line 8616\.
> Debugged program terminated\.  Use q to quit or R to restart\,
> use o inhibit\_exit to avoid stopping after program termination\,
> h q\, h R or h o to get additional info\.
>   DB\<1> q
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 

I can reproduce this behavior in perl\-5\.20 and \-5\.22 \-\- but it appears to have been fixed in \-5\.24\.

5\.20
\#\#\#\#\#
\[perl\] 581 $ perl \-v | head \-2 | tail \-1
This is perl 5\, version 20\, subversion 3 \(v5\.20\.3\) built for x86\_64\-linux
\[perl\] 582 $ perl \-e 'die\(\{\}\);' || echo
HASH\(0x273c148\)
\[perl\] 583 $ perl \-de 'die\(\{\}\);' || echo

Loading DB routines from perl5db\.pl version 1\.44\_01
Editor support available\.

Enter h or 'h h' for help\, or 'man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:    die\(\{\}\);
&nbsp;  DB\<1> c
1 at /home/jkeenan/perl5/perlbrew/perls/perl\-5\.20\.3/lib/5\.20\.3/perl5db\.pl line 8616\.
Debugged program terminated\.  Use q to quit or R to restart\,
use o inhibit\_exit to avoid stopping after program termination\,
h q\, h R or h o to get additional info\.
&nbsp;  DB\<1> q
\#\#\#\#\#

5\.22
\#\#\#\#\#
\[perl\] 586 $ perl \-v | head \-2 | tail \-1
This is perl 5\, version 22\, subversion 3 \(v5\.22\.3\) built for x86\_64\-linux
\[perl\] 587 $ perl \-e 'die\(\{\}\);' || echo
HASH\(0x140d160\)
\[perl\] 588 $ perl \-de 'die\(\{\}\);' || echo

Loading DB routines from perl5db\.pl version 1\.49\_001
Editor support available\.

Enter h or 'h h' for help\, or 'man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:    die\(\{\}\);
&nbsp;  DB\<1> c
1 at /home/jkeenan/perl5/perlbrew/perls/perl\-5\.22\.3/lib/5\.22\.3/perl5db\.pl line 8674\.
Debugged program terminated\.  Use q to quit or R to restart\,
use o inhibit\_exit to avoid stopping after program termination\,
h q\, h R or h o to get additional info\.
&nbsp;  DB\<1> q
\#\#\#\#\#

5\.24
\#\#\#\#\#
\[perl\] 590 $ perl \-v | head \-2 | tail \-1
This is perl 5\, version 24\, subversion 1 \(v5\.24\.1\) built for x86\_64\-linux
\[perl\] 591 $ perl \-e 'die\(\{\}\);' || echo
HASH\(0x19e9178\)
\[perl\] 592 $ perl \-de 'die\(\{\}\);' || echo

Loading DB routines from perl5db\.pl version 1\.49\_05
Editor support available\.

Enter h or 'h h' for help\, or 'man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:    die\(\{\}\);
&nbsp;  DB\<1> c
HASH\(0x17483a0\)Debugged program terminated\.  Use q to quit or R to restart\,
use o inhibit\_exit to avoid stopping after program termination\,
h q\, h R or h o to get additional info\.
&nbsp;  DB\<1> q
\#\#\#\#\#




> This behavior is because perl5db\.pl is calling Carp&#8203;::longmess in
> scalar context
> instead of list context\. When the first argument of Carp&#8203;::longmess \(or
> shortmess\) is a reference\, it returns it's argument list \(@&#8203;\_\)
> unmodified\.
> 
> I've attached a patch \(debug\_exception\_1\.patch\) that calls
> Carp&#8203;::longmess in
> list context to provide more useful messages\.

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @Smylers

James E Keenan via RT writes&#8203;:

> On Wed\, 01 Nov 2017 01&#8203;:03&#8203;:53 GMT\, jkahrman@&#8203;mathworks\.com wrote&#8203;: 
> 
> > If an exception object is thrown while stepping through the
> > debugger\, the debugger prints '1 at perl5db\.pl line 8616'
> 
> I can reproduce this behavior in perl\-5\.20 and \-5\.22 \-\- but it appears
> to have been fixed in \-5\.24\.

Fixed by 6ca94a7e\, which changes Carp \(specifically changing what
longmess can return in scalar context\)\, making the debugger's use of
longmess work\.

That became Carp 1\.39\. The most recent standalone Carp on Cpan is 1\.38\.

Smylers

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
